doc/proposals/helm-operator.md

```diff
@@ -0,0 +1,127 @@
+
+* /operator-sdk/pkg/helm/controller
+  * Will contain the Helm controller.
+  * Will contain an exposed reconciler. The default `Add` method will use this reconciler.
+
+* /operator-sdk/pkg/helm/engine
+  * Will contain a Helm Engine implementation that adds owner references to generated Kubernetes resource assets, which is necessary for garbage collection of Helm chart resources.
+
+* /operator-sdk/pkg/helm/internal
+  * Will contain types and utilities used by other Helm packages in the SDK.
+
+* /operator-sdk/pkg/helm/release
+  * Will contain the Manager types and interfaces. A Manager is responsible for:
+    * Implementing Helm's Tiller functions that are necessary to install, update, and uninstall releases.
+    * Reconciling an existing release's resources.
+  * A default Manager implementation is provided in this package but is not exported.
+  * Package functions:
+    * NewManager - method that returns a new Manager for a provided helm chart.
+    * NewManagersFromEnv - method that returns a map of GVK to Manager types based on environment variables.
+    * NewManagersFromFile - method that returns a map of GVK to Manager types based on a provided config file.
+
+### Commands
+
+We are adding and updating existing commands to accommodate the helm operator.  Changes to the `cmd` package as well as changes to the generator are needed.
+
+#### New
+
+New functionality will be updates to allow Helm operator developers to create a new boilerplate operator structure with everything necessary to get started developing and deploying a Helm operator with the SDK.
+
+```
+operator-sdk new <project-name> --type=helm --kind=<kind> --api-version=<group/version>
+```
+
+This will be new scaffolding for the above command under the hood. We will:
+* Create a `./<project-name>` directory.
+* Create a `./<project-name>/charts` directory.
+* Generate a simple default chart at `./<project-name>/charts/<kind>`.
+* Create a new watches file at `./<project-name>/watches.yaml`. The chart and GVK will be defaulted based on input to the `new` command.
+* Create a `./<project-name>/deploy` with the Kubernetes resource files necessary to run the operator.
+* Create a `./build/Dockerfile` that uses the watches file and the helm chart. It will use the Helm operator as its base image.
+
+The resulting structure will be:
+
+```
+<project-name>
+|   watches.yaml
+|
+|-- build
+|   |   Dockerfile
+|
+|-- charts
+|   |-- <kind>
+|       |   Chart.yaml
+|       |   ...
+|
+|-- deploy
+|   |   operator.yaml
+|   |   role_binding.yaml
+|   |   role.yaml
+|   |   service_account.yaml
+|   |
+|   |-- crds
+|       |   <gvk>_crd.yaml
+|       |   <gvk>_cr.yaml
+```
+
+#### Add
+
+Add functionality will be updated to allow Helm operator developers to add new CRDs/CRs and to update the watches.yaml file for additional Helm charts. The command helps when a user wants to watch more than one CRD for their operator.
```

Comment:
Is this is only for creating CRDs? I think you may just want to re-use the ansible operator stuff for just ‘add crd’.

Reply:
Ahh. I didn't see the `add crd` subcommand the first time around. 

My initial thought was that `app api` would be operator type aware.
* For `go`, it would do what it does now.
* For `ansible`, it would create the crds, update the watches file, and scaffold out a `new roles/<Kind>` Ansible role.
* For `helm` it would create the crds, update the watches file, and scaffold out a new `helm-charts/<Kind>` Helm chart.

If that was the pattern, that could also potentially align how `operator-sdk new` works (i.e. --api-version and --kind would no longer be valid for a new operator of any type)

I was surprised when I ran `operator-sdk add api` in an ansible operator project and it spit out Go files.

Reply:
If someone wants to convert to a golang operator from an ansible operator, don't want to lock them in. 

I also don't like calling this adding an API, IMO an API is what this does adds the go structures for the API. I would prefer to keep the CRD command as I think it is much more descriptive. When discussing the ansible operator, we never talk about adding API's or watching API's we discuss CRDs and CRs.

I also do not want new for ansible operator does not allow `--api-version` and `--kind`. This is really important to the ansible operator workflow.

Reply:
Yeah the command `add api` is only valid in the context of a Go operator.
We can use the command [`add crd`](https://github.com/operator-framework/operator-sdk/blob/master/commands/operator-sdk/cmd/add/crd.go#L34-L44) here.

> I was surprised when I ran operator-sdk add api in an ansible operator project and it spit out Go files.

We need to make sure to check not run the commands specific to the Go type only run in a Go operator project:
- `add api`
- `add controller`
- `generate k8s`

I'll create an issue to fix this.

Reply:
I don't think we scaffold a new role for `add crd` although it kind of makes sense that you would want to associate a new role/chart with a new type.
But then do we update watches.yaml file as well to point to the new chart/role?
And if we do couple it with the `add crd` command then we should ensure that it handles Go/Ansible/Helm project types differently.

The only downside is that this couples a bunch of different things to the `add crd` command.

Reply:
@hasbro17 Is there a use case for `add crd` that wouldn't involve other changes? Along those same lines, for a Go project, when would one use `add crd` and not `add api`?

If there is a use case for running `add crd` standalone, I can see how coupling it to these other things may not be desirable.

But to me, it seems like the user is always going to follow `add crd` up with other manual changes that will likely depend on the type of the operator. For `ansible` and `helm`, I think it's pretty likely users will be manually adding a new role/chart and updating watches.yaml.

If users are doing that manually, it increases the likelihood that they forget to change something somewhere or make a typo that causes their new CRD to fail in some way.

Reply:
@joelanford `add crd` was initially added for the Ansible project.
I don't think there was a use case for it in a Go project, since you almost always need the `pkg/apis/...` files to register the CRD type with the scheme so that the operator can actually watch or CRUD that type.

Maybe if you're dealing with unstructured types(like the ansible and helm operator do) then you don't need the api definition `pkg/apis/...`. But I don't think that's a common use case.

So yeah I think running `add crd` standalone isn't necessary for the Go project when we have `add api`. 
So disregarding the Go project we could couple it with adding a role/chart and updating watches.yaml. 
@shawn-hurley WDYT?  Leave it up to the users. Couple it with `add crd`. Or a new subcommand.